fuzzers/060-bram-cascades/top.py

```diff
@@ -0,0 +1,221 @@
+        net = 'lut_{}_o'.format(self.output_lut_idx)
+        self.output_lut_idx += 1
+        return net
+
+    def create_wires_and_luts(self):
+        if self.output_lut_idx > self.input_lut_idx:
+            nluts = self.output_lut_idx
+        else:
+            nluts = self.input_lut_idx
+            if self.lut_input_idx > 0:
+                nluts += 1
+
+        for lut in range(nluts):
+            yield """
+            wire [5:0] lut_{lut}_i;
+            wire lut_{lut}_o;
+
+            (* KEEP, DONT_TOUCH *)
+            LUT6 lut_{lut} (
+                .I0(lut_{lut}_i[0]),
+                .I1(lut_{lut}_i[1]),
+                .I2(lut_{lut}_i[2]),
+                .I3(lut_{lut}_i[3]),
+                .I4(lut_{lut}_i[4]),
+                .I5(lut_{lut}_i[5]),
+                .O(lut_{lut}_o)
+            );
+            """.format(lut=lut)
+
+
+def sdp_bram(name, width, address_bits):
+    depth = 2**address_bits
+
+    return '''
+module {name}(
+    // Write port
+    input wrclk,
+    input [{width}-1:0] di,
+    input wren,
+    input [{address_bits}-1:0] wraddr,
+    // Read port
+    input rdclk,
+    input rden,
+    input [{address_bits}-1:0] rdaddr,
+    output reg [{width}-1:0] do);
+
+    (* ram_style = "block" *) reg [{width}-1:0] ram[0:{depth}];
+
+    always @ (posedge wrclk) begin
+        if(wren == 1) begin
+            ram[wraddr] <= di;
+        end
+    end
+
+    always @ (posedge rdclk) begin
+        if(rden == 1) begin
+            do <= ram[rdaddr];
+        end
+    end
+
+endmodule
+    '''.format(
+        name=name,
+        width=width,
+        address_bits=address_bits,
+        depth=depth,
+    )
+
+
+MAX_BRAM = 8
+
+
+def emit_sdp_bram(luts, name, modules, lines, width, address_bits):
+    modules.append(sdp_bram(name=name, width=width, address_bits=address_bits))
+
+    lines.append(
+        '''
+    wire [{address_bits}-1:0] {name}_wraddr;
+    wire [{address_bits}-1:0] {name}_rdaddr;
+    '''.format(
+            name=name,
+            address_bits=address_bits,
+        ))
+
+    for bit in range(address_bits):
+        lines.append(
+            """
+    assign {name}_wraddr[{bit}] = {net};""".format(
+                name=name, bit=bit, net=luts.get_next_output_net()))
+
+    for bit in range(address_bits):
+        lines.append(
+            """
+    assign {name}_rdaddr[{bit}] = {net};""".format(
+                name=name, bit=bit, net=luts.get_next_output_net()))
+
+    lines.append(
+        '''
+    (* KEEP, DONT_TOUCH *)
+    {name} {name}_inst(
+        .wraddr({name}_wraddr),
+        .rdaddr({name}_rdaddr)
+    );
+    '''.format(name=name))
+
+    return width, address_bits, math.ceil(
+        float(width) / 72) * 72 * (2**address_bits)
+
+
+def max_address_bits(width):
+    return math.floor(math.log(float(MAX_BRAM * 36 * 1024) / width, 2))
+
+
+def random_sdp_bram(luts, name, modules, lines):
+    sdp_choices = set()
+
+    for width in (1, 2, 4, 8, 16, 18, 32, 36):
```

Comment:
SDP has width 72 i think?

Reply:
The loop below starts to gets to those values